### PR TITLE
LRScheduledTFModel: Super Convergence and Learning rate Schedulers

### PR DESCRIPTION
**Pull request by vikmary**
- [x] add learning rate & momentum schedulers
- [x] wrap TFModel with EnhancedTFModel that supports schedulers
- [x] rewrite go_bot, ner and squad to inherit from EnhancedModel
- [x] refactor go_bot
- [x] Super Convergence tutorial

### CODE REVIEW

File: deeppavlov/core/commands/train.py [hunk 1]
```diff
@@ -83,21 +83,43 @@ def fit_chainer(config: dict, iterator: Union[DataLearningIterator, DataFittingI
...
+                for name, scores in result.items():
+                    for i, score in enumerate(scores):
+                        metric_sum = tf.Summary(value=[tf.Summary.Value(tag='fit/' + name,
+                                                                        simple_value=score), ])
+                        writer.add_summary(metric_sum, i)
+
             component.save()
 
-        if 'fit_on_batch' in component_config:
+        if 'fit_on_batches' in component_config:
```
Comment (yoptar) [inline deeppavlov/core/commands/train.py]: it would be great if both options would work, but for an older one a deprecation warning would be thrown

File: deeppavlov/core/models/tf_model.py [hunk 2]
```diff
@@ -133,3 +147,341 @@ def print_number_of_parameters():
...
+                # constant end_val
+                return self.end_val
+            else:
+                # raising from start_val to end_val for 1/10 * nb steps
+                pct = 10.0 * self.iters / self.nb
+                return self.start_val * (1 + pct * (self.div - 1))
+
+
+class EnhancedTFModel(TFModel):
+    """TFModel anhanced with optimizer, learning rate and momentum configuration"""
```
Comment (yoptar) [inline deeppavlov/core/models/tf_model.py]:
```suggestion
    """TFModel enhanced with optimizer, learning rate and momentum configuration"""
```

File: deeppavlov/core/common/chainer.py [hunk 3]
```diff
@@ -192,3 +192,4 @@ def destroy(self):
         for in_params, out_params, component in self.pipe:
             if inspect.ismethod(getattr(component, 'destroy', None)):
                 component.destroy()
+            del component
```
Comment (yoptar) [inline deeppavlov/core/common/chainer.py]: This does nothing. `component` variable is deleted in every iteration of this `for`. And the last one is deleted after leaving the `destroy` method

Comment (vikmary) [reply deeppavlov/core/common/chainer.py]: Why `component` variable is deleted without `del component`?

File: deeppavlov/core/models/lr_scheduled_tf_model.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+
+class LRScheduledTFModel(TFModel):
+    """
+    TFModel enhanced with optimizer, learning rate and momentum
+    management and search.
+    """
+    def __init__(self,
+                 learning_rate: Union[float, Tuple[float, float]] = None,
+                 learning_rate_decay: Union[str, DecayType, List[Any]] = DecayType.NO,
```
[the remark this answers is not in the document]
Comment (vikmary) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: :+1: Тогда только Union[str, DecayType, Tuple[Union[str,DecayType],Any]]

File: deeppavlov/core/models/lr_scheduled_tf_model.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+            self.div = 1.0 if not self.start_val else self.end_val / self.start_val
+        if self.dec_type == DecayType.TRAPEZOID:
+            self.div = 1.0 if not self.start_val else self.end_val / self.start_val
+
+    def __str__(self):
+        return f"DecayScheduler(start_val={self.start_val}, end_val={self.end_val}"\
+            f", dec_type={self.dec_type.name}, num_it={self.nb}, extra={self.extra})"
+
+    def next_val(self):
+        self.iters = min(self.iters + 1, self.nb)
```
[the remark this answers is not in the document]
Comment (vikmary) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: Right.

[the remark this answers is not in the document]
Comment (vikmary) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: Is it unexpected behavior?

File: deeppavlov/core/models/lr_scheduled_tf_model.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+        # print(f"iters = {self.iters}/{self.nb}")
+        if self.dec_type == DecayType.NO:
+            return self.start_val
+        elif self.dec_type == DecayType.LINEAR:
+            pct = self.iters / self.nb
+            return self.start_val + pct * (self.end_val - self.start_val)
+        elif self.dec_type == DecayType.COSINE:
+            cos_out = math.cos(math.pi * self.iters / self.nb) + 1
+            return self.end_val + (self.start_val - self.end_val) / 2 * cos_out
```
[the remark this answers is not in the document]
Comment (vikmary) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: Wow, forgot all about it. Should have added to tutorial. Seems like a cyclic lr schedule?

File: deeppavlov/core/models/lr_scheduled_tf_model.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+        self._external_mom = False
+
+    def load(self, exclude_scopes: Optional[Iterable] = ('Optimizer',
+                                                         'learning_rate',
+                                                         'momentum')):
+        return super().load(exclude_scopes=exclude_scopes)
+
+    def fit(self, *args):
+        data = list(zip(*args))
+        self.save()
```
[the remark this answers is not in the document]
Comment (vikmary) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: Just for convenience.
If we intended to load model from saved and continue training, then we would like to fit learning rate on one epoch, and then load from saved one more time.
But if we didn't have saved model from the start, then after fitting would like to initialize with zeros and do not know how to do this (special method initialize_from_scratch() would be an ugly solution).

File: deeppavlov/core/models/lr_scheduled_tf_model.py [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+        if self._fit_batch_size is None:
+            raise ConfigError("in order to use fit() method"
+                              " set `fit_batch_size` parameter")
+        bs = int(self._fit_batch_size)
+        data_len = len(data)
+        num_batches = self._fit_max_batches or ((data_len - 1) // bs + 1)
+
+        avg_loss = 0.
```
[the remark this answers is not in the document]
Comment (vikmary) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: No, because there is a bias correction implementation.
See https://github.com/deepmipt/DeepPavlov/blob/dd0c79da51d81afc3a04740b7364f87740c27a07/deeppavlov/core/models/lr_scheduled_tf_model.py#L248-L249
When `avg_loss` equals `0`, after 248th line it will be equal to `(1 - self._fit_beta)*report['loss']`.
249th line removes the `(1 - self._fit_beta)` multiplier before `report['loss']`

File: deeppavlov/core/models/lr_scheduled_tf_model.py [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+            self._update_tf_variables(learning_rate=self._lr)
+
+            if i >= num_batches:
+                break
+        # best_lr /= 10
+        end_val = self._get_best(lrs, losses)
+
+        start_val = end_val
+        if self._lr_schedule.dec_type in (DecayType.ONECYCLE, DecayType.TRAPEZOID):
+            start_val = end_val / self._fit_learning_rate_div
```
[the remark this answers is not in the document]
Comment (vikmary) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: `_fit_learning_rate_div` is only used in the main training phase. What do you think it should be used for?

File: deeppavlov/core/models/lr_scheduled_tf_model.py [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+        self._mom = self._mom_schedule.start_val
+        self._update_tf_variables(learning_rate=self._lr, momentum=self._mom)
+        return {'smoothed_loss': losses, 'learning_rate': lrs}
+
+    @staticmethod
+    def _get_best(values, losses, max_loss_div=0.9, min_val_div=10.0):
+        assert len(values) == len(losses), "lengths of values and losses should be equal"
+        min_ind = np.argmin(losses)
+        for i in range(min_ind - 1, 0, -1):
+            if (losses[i] * max_loss_div > losses[min_ind]) or\
```
[the remark this answers is not in the document]
Comment (vikmary) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: True

File: deeppavlov/configs/ner/ner_ontonotes.json [hunk 11]
```diff
@@ -106,6 +106,7 @@
         "intra_layer_dropout": false,
         "l2_reg": 0,
         "learning_rate": 3e-3,
+        "learning_rate_drop_patience": 3,
```
[the remark this answers is not in the document]
Comment (vikmary) [reply deeppavlov/configs/ner/ner_ontonotes.json]: I haven't checked. All ner configs are changed so that old learning rate schedule is preserved.

File: deeppavlov/core/models/lr_scheduled_tf_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mu-arkhipov) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: No, it's fine

File: deeppavlov/core/models/lr_scheduled_tf_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mu-arkhipov) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: let the current `report['loss']` be equal to 1, then you will write `(1 - 0.98) * 1 = 0.02` to the `avg_loss`. This means that you are starting from the 0 and to reach 1 you will need a lot of steps.

File: deeppavlov/core/models/lr_scheduled_tf_model.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mu-arkhipov) [reply deeppavlov/core/models/lr_scheduled_tf_model.py]: My bad, may be the name is a little bit confusing
